skglm/solvers/prox_newton_solver.py

```diff
@@ -0,0 +1,312 @@
+import warnings
+import numpy as np
+from scipy import sparse
+from numba import njit
+from skglm.datafits import Logistic, Logistic_32
+from skglm.solvers.common import construct_grad, construct_grad_sparse, dist_fix_point
+from skglm.utils import (
+    AndersonAcceleration, weighted_dot, weighted_dot_sparse, xj_dot_sparse, sigmoid)
+
+
+def prox_newton_solver(
+    X, y, datafit, penalty, w, Xw, max_iter=50, max_epochs=1000, max_backtrack=10,
+    min_pn_cd_epochs=2, max_pn_cd_epochs=10, p0=10, tol=1e-4, ws_strategy="subdiff",
+    verbose=0
+):
+    r"""Run a prox-Newton solver.
+
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features)
+        Training data.
+
+    y : array, shape (n_samples,)
+        Target values.
+
+    datafit : instance of Datafit class
+        Datafitting term.
+
+    penalty : instance of Penalty class
+        Penalty used in the model.
+
+    w : array, shape (n_features,)
+        Coefficient vector.
+
+    Xw : array, shape (n_samples,)
+        Model fit.
+
+    max_iter : int, optional
+        Maximum number of iterations (definition of working set and
+        resolution of problem restricted to features in working set).
+
+    max_epochs : int, optional
+        Maximum number of prox-Newton epochs on each subproblem.
+
+    max_backtrack : int, optinal
+        Maximum number of backtracking steps for the line search.
+
+    min_pn_cd_epochs : int, optional
+        Minimum number of iterations used in the prox-Newton coordinate
+        descent.
+
+    max_pn_cd_epochs : int, optional
+        Maximum number of iterations used in the prox-Newton coordinate
+        descent.
+
+    p0 : int, optional
+        First working set size.
+
+    tol : float, optional
+        The tolerance for the optimization.
+
+    ws_strategy : ('subdiff'|'fixpoint'), optional
+        The score used to build the working set.
+
+    verbose : bool or int, optional
+        Amount of verbosity. 0/False is silent.
+
+    Returns
+    -------
+    w : array, shape (n_features,)
+        Coefficient vector.
+
+    obj_out : array, shape (n_iter,)
+        The objective values at every outer iteration.
+
+    stop_crit : float
+        Value of stopping criterion at convergence.
+    """
+    if ws_strategy not in ("subdiff", "fixpoint"):
+        raise ValueError(f"Unsupported value for ws_strategy: {ws_strategy}")
+    if not isinstance(datafit, (Logistic, Logistic_32)):
+        raise ValueError("Prox-Newton solver only supports Logistic datafits.")
+    n_features = X.shape[1]
+    if n_features >= 10_000:
+        warnings.warn(
+            "Prox-Newton solver can be prohibitively slow for high-dimensional " +
```

Comment:
not sure this warning should be raised : it is coupled with a working set method so the number of features is not really relevant here